core/shared/src/main/scala/scalaz/zio/stm/STM.scala

```diff
@@ -377,22 +407,93 @@ object STM {
     final def addTodo(txnId: Long, journal: Journal, todoEffect: Todo): Unit = {
       val it = journal.entrySet().iterator()
 
-      while (it.hasNext()) {
+      while (it.hasNext) {
         val tref = it.next().getValue().tref
 
         var loop = true
         while (loop) {
           val oldTodo = tref.todo.get
 
-          val newTodo = oldTodo updated (txnId, todoEffect)
+          val newTodo = oldTodo.updated(txnId, todoEffect)
 
           loop = !tref.todo.compareAndSet(oldTodo, newTodo)
         }
       }
     }
 
-    final val succeedUnit: TRez[Nothing, Unit] =
-      TRez.Succeed(())
+    /**
+     * Runs all the todos.
+     */
+    def completeTodos[E, A](io: IO[E, A], journal: Journal, platform: Platform): TryCommit[E, A] = {
+      val todos = collectTodos(journal)
+
+      if (todos.size > 0) platform.executor.submitOrThrow(() => execTodos(todos))
+
+      TryCommit.Done(io)
+    }
+
+    def tryCommitAsync[E, A](journal: Journal, platform: Platform, stm: STM[E, A], txnId: Long, done: AtomicBoolean)(
+      k: IO[E, A] => Unit
+    ): Unit = {
+      def complete(io: IO[E, A]): Unit = { done.set(true); k(io) }
+      def suspend(journal: Journal): Unit = {
+        addTodo(txnId, journal, () => tryCommitAsync(null, platform, stm, txnId, done)(k))
+
+        if (isInvalid(journal)) tryCommit(platform, stm) match {
```

Comment:
Everything is super clean, but I don't understand this line. Isn't it redundant with the line above?

Reply:
In between the time when the STM was executed and the "todo" was added, it may happen that the variables underlying the transaction have changed (by some _other_ transaction), and the "todo" lists cleared by that transaction. In this case, the suspended transaction would never be retried. This line detects this case, and re-runs the transaction again if necessary.

This way of structuring the code allows dramatic minimization of the commit lock, albeit at the cost of trickier reasoning about correctness.